File: python/tournament/tournament.py
class Team:

    POINTS_PER_WIN = 3
    POINTS_PER_DRAW = 1

    def __init__(self, name):
        self.name = name
        self.wins = 0
        self.draws = 0
        self.losses = 0

    @property
    def matches_played(self):
        return self.wins + self.losses + self.draws

    @property
    def points(self):
        return (self.wins * self.POINTS_PER_WIN) + (self.draws * self.POINTS_PER_DRAW)

    def tally_win(self):
        self.wins += 1

    def tally_draw(self):
        self.draws += 1

    def tally_loss(self):
        self.losses += 1

    def __str__(self):
        return '{:<30} | {:^3}| {:^3}| {:^3}| {:^3}| {:>2}'.format(self.name, self.matches_played, self.wins, self.draws, self.losses, self.points)


class Tournament:

    WIN = 'win'
    DRAW = 'draw'
    LOSS = 'loss'
    RESULT_SEPERATOR = ';'
    COLUMN_HEADERS = ['Team', 'MP', 'W', 'D', 'L', 'P']

    def __init__(self, results):
        self._teams = {}
        if results:
            self.parse(results)

    def results_table(self):
        table = [self.table_header()]
        for team in self.sorted_teams():
            table.append(str(team))
        return "\n".join(table)

    def sorted_teams(self):
        alphabetic = sorted(self._teams.values(), key=lambda team: team.name)
        alphabetic_descending_points = sorted(alphabetic, key=lambda team: team.points, reverse=True)
        return alphabetic_descending_points

    def parse(self, results):
        for result in results.split("\n"):
            team_a, team_b, outcome = result.split(self.RESULT_SEPERATOR)
            self.maybe_initialize_teams(team_a, team_b)
            self.tally_outcome(team_a, team_b, outcome)

    def tally_outcome(self, team_a, team_b, outcome):
        if outcome == self.WIN:
            self.tally_win(team_a, team_b)
        if outcome == self.LOSS:
            self.tally_loss(team_a, team_b)
        if outcome == self.DRAW:
            self.tally_draw(team_a, team_b)

    def tally_win(self, winner, loser):
        self._teams[winner].tally_win()
        self._teams[loser].tally_loss()

    def tally_loss(self, loser, winner):
        self._teams[loser].tally_loss()
        self._teams[winner].tally_win()

    def tally_draw(self, team_a, team_b):
        self._teams[team_a].tally_draw()
        self._teams[team_b].tally_draw()

    def maybe_initialize_teams(self, team_a, team_b):
        self.maybe_initialize_team(team_a)
        self.maybe_initialize_team(team_b)

    def maybe_initialize_team(self, name):
        if name not in self._teams:
            self._teams[name] = Team(name)

    def table_header(self):
        return '{:<30} | {:^3}| {:^3}| {:^3}| {:^3}| {:>2}'.format(*self.COLUMN_HEADERS)

def tally(results):
    return Tournament(results).results_table()

File: go/tournament/tournament.go
package tournament

import (
	"fmt"
	"io"
	"sort"
	"strings"
)

type stat struct {
	Name          string
	MatchesPlayed int
	Wins          int
	Draws         int
	Loses         int
	Points        int
}

func Tally(r io.Reader, w io.Writer) error {
	lines, err := getLines(r)
	if err != nil {
		return err
	}

	teamsToScores := map[string]*stat{}
	for _, line := range lines {
		if isComment(line) || isEmptyLine(line) {
			continue
		}

		a, b, outcome, err := parseLine(line)
		if err != nil {
			return err
		}

		teamsToScores = initializeTeamIfNotExists(teamsToScores, a)
		teamsToScores = initializeTeamIfNotExists(teamsToScores, b)

		switch outcome {
		case "win":
			handleWin(teamsToScores, a, b)
		case "loss":
			handleWin(teamsToScores, b, a)
		case "draw":
			handleDraw(teamsToScores, a, b)
		default:
			return fmt.Errorf("unexpected outcome %s", outcome)
		}
	}

	io.WriteString(w, getTable(teamsToScores))
	return nil
}

func getTable(teamsToScores map[string]*stat) (result string) {
	sorted := sortTeamsByPointsThenAlphabetically(teamsToScores)
	var header = fmt.Sprintf("%-30s |%3s |%3s |%3s |%3s |%3s\n", "Team", "MP", "W", "D", "L", "P")
	result += header
	for _, stat := range sorted {
		result += fmt.Sprintf("%-30s |%3d |%3d |%3d |%3d |%3d\n", stat.Name, stat.MatchesPlayed, stat.Wins, stat.Draws, stat.Loses, stat.Points)
	}
	return result
}

func sortTeamsByPointsThenAlphabetically(teamsToScores map[string]*stat) (result []*stat) {
	for _, stat := range teamsToScores {
		result = append(result, stat)
	}
	sort.SliceStable(result, func(i, j int) bool {
		if result[i].Points != result[j].Points {
			return result[i].Points > result[j].Points
		}
		return result[i].Name < result[j].Name
	})

	return result
}

func isComment(line string) bool {
	return strings.HasPrefix(line, "#")
}

func isEmptyLine(line string) bool {
	return strings.TrimSpace(line) == ""
}

func getLines(r io.Reader) ([]string, error) {
	input, err := io.ReadAll(r)
	if err != nil {
		return []string{}, err
	}
	lines := strings.Split(string(input), "\n")
	return lines, nil
}

func initializeTeamIfNotExists(teamsToScores map[string]*stat, team string) map[string]*stat {
	if _, ok := teamsToScores[team]; !ok {
		teamsToScores[team] = &stat{
			Name: team,
		}
	}
	return teamsToScores
}

func parseLine(line string) (a string, b string, outcome string, err error) {
	parts := strings.Split(line, ";")
	if len(parts) < 3 {
		return "", "", "", fmt.Errorf("unexpected number of parts in line. Expected 3 parts delimited by semicolon. Got %d parts", len(parts))
	}
	return parts[0], parts[1], parts[2], nil
}

func handleWin(teamsToScores map[string]*stat, winner string, loser string) {
	teamsToScores[winner].Wins += 1
	teamsToScores[winner].Points += 3
	teamsToScores[loser].Loses += 1
	teamsToScores[winner].MatchesPlayed += 1
	teamsToScores[loser].MatchesPlayed += 1
}

func handleDraw(teamsToScores map[string]*stat, a string, b string) {
	teamsToScores[a].Draws += 1
	teamsToScores[b].Draws += 1
	teamsToScores[a].Points += 1
	teamsToScores[b].Points += 1
	teamsToScores[a].MatchesPlayed += 1
	teamsToScores[b].MatchesPlayed += 1
}
